File: training_data/reflection-tuning/day22_part1_2022.py
def parse_input(filename):
    with open(filename, 'r') as f:
        map_data, path = f.read().split('\n\n')
    
    map_lines = map_data.split('\n')
    max_width = max(len(line) for line in map_lines)
    map_grid = [list(line.ljust(max_width)) for line in map_lines]
    
    return map_grid, path.strip()

def find_start(map_grid):
    return (0, map_grid[0].index('.'))

def move(map_grid, pos, direction):
    directions = [(0, 1), (1, 0), (0, -1), (-1, 0)]  # R, D, L, U
    row, col = pos
    dr, dc = directions[direction]
    new_row, new_col = (row + dr) % len(map_grid), (col + dc) % len(map_grid[0])
    
    while map_grid[new_row][new_col] == ' ':
        new_row, new_col = (new_row + dr) % len(map_grid), (new_col + dc) % len(map_grid[0])
    
    if map_grid[new_row][new_col] == '#':
        return pos
    return (new_row, new_col)

def solve(map_grid, path):
    pos = find_start(map_grid)
    direction = 0
    
    steps = ''
    for char in path + 'X':  # Add dummy char to process last number
        if char.isdigit():
            steps += char
        else:
            if steps:
                for _ in range(int(steps)):
                    pos = move(map_grid, pos, direction)
                steps = ''
            if char == 'R':
                direction = (direction + 1) % 4
            elif char == 'L':
                direction = (direction - 1) % 4
    
    row, col = pos
    return 1000 * (row + 1) + 4 * (col + 1) + direction

map_grid, path = parse_input("input.txt")
result = solve(map_grid, path)
print(result)

File: swift/day22_part1_2022.swift

import Foundation

// Read input from file
guard let input = try? String(contentsOfFile: "input.txt") else {
    fatalError("Could not read input file")
}

// Parse input into map and path
let parts = input.components(separatedBy: "\n\n")
let mapLines = parts[0].split(separator: "\n").map { Array(String($0)) }
let pathString = parts[1].trimmingCharacters(in: .whitespacesAndNewlines)

// Determine map dimensions and pad with spaces for easier wrapping
let maxRowLength = mapLines.map { $0.count }.max()!
let map: [[Character]] = mapLines.map { row in
    var paddedRow = row
    while paddedRow.count < maxRowLength {
        paddedRow.append(" ")
    }
    return paddedRow
}

// Parse path into instructions
var instructions: [(Int, Character?)] = []
var currentNumber = ""
for char in pathString {
    if char.isNumber {
        currentNumber.append(char)
    } else {
        instructions.append((Int(currentNumber)!, char))
        currentNumber = ""
    }
}
if !currentNumber.isEmpty {
    instructions.append((Int(currentNumber)!, nil))
}

// Define directions and initial position
let directions: [(Int, Int)] = [(0, 1), (1, 0), (0, -1), (-1, 0)] // R, D, L, U
var row = 0
var col = map[0].firstIndex(of: ".")!
var facing = 0

// Function to find the next valid position after wrapping
func findWrappedPosition(row: Int, col: Int, direction: (Int, Int)) -> (Int, Int) {
    var newRow = row
    var newCol = col
    while true {
        newRow = (newRow + direction.0 + map.count) % map.count
        newCol = (newCol + direction.1 + maxRowLength) % maxRowLength
        if map[newRow][newCol] != " " {
            return (newRow, newCol)
        }
    }
}

// Follow the path
for (steps, turn) in instructions {
    for _ in 0..<steps {
        let nextRow = (row + directions[facing].0 + map.count) % map.count
        let nextCol = (col + directions[facing].1 + maxRowLength) % maxRowLength
        
        if map[nextRow][nextCol] == "#" {
            break // Hit a wall
        } else if map[nextRow][nextCol] == "." {
            row = nextRow
            col = nextCol
        } else { // Wrap around
            let (wrappedRow, wrappedCol) = findWrappedPosition(row: row, col: col, direction: directions[facing])
            if map[wrappedRow][wrappedCol] == "#" {
                break // Hit a wall after wrapping
            } else {
                row = wrappedRow
                col = wrappedCol
            }
        }
    }
    
    if let turn = turn {
        if turn == "R" {
            facing = (facing + 1) % 4
        } else {
            facing = (facing + 3) % 4
        }
    }
}

// Calculate final password
let finalPassword = 1000 * (row + 1) + 4 * (col + 1) + facing
print(finalPassword)
